nupic/research/frameworks/greedy_infomax/utils/model_utils.py

```diff
@@ -0,0 +1,335 @@
+    model,
+    loader,
+    optimizer,
+    device,
+    criterion=F.nll_loss,
+    complexity_loss_fn=None,
+    batches_in_epoch=sys.maxsize,
+    active_classes=None,
+    pre_batch_callback=None,
+    post_batch_callback=None,
+    transform_to_device_fn=None,
+    progress_bar=None,
+):
+    """Train the given model by iterating through mini batches. An epoch ends
+    after one pass through the training set, or if the number of mini batches
+    exceeds the parameter "batches_in_epoch".
+
+    :param model: pytorch model to be trained
+    :type model: torch.nn.Module
+    :param loader: train dataset loader
+    :type loader: :class:`torch.utils.data.DataLoader`
+    :param optimizer: Optimizer object used to train the model.
+           This function will train the model on every batch using this optimizer
+           and the :func:`torch.nn.functional.nll_loss` function
+    :param device: device to use ('cpu' or 'cuda')
+    :type device: :class:`torch.device
+    :param criterion: loss function to use
+    :type criterion: function
+    :param complexity_loss_fn: a regularization term for the loss function
+    :type complexity_loss_fn: function
+    :param batches_in_epoch: Max number of mini batches to test on
+    :type batches_in_epoch: int
+    :param active_classes: a list of indices of the heads that are active for a given
+                           task; only relevant if this function is being used in a
+                           continual learning scenario
+    :type active_classes: list of int or None
+    :param pre_batch_callback: Callback function to be called before every batch
+                               with the following parameters: model, batch_idx
+    :type pre_batch_callback: function
+    :param post_batch_callback: Callback function to be called after every batch
+                                with the following parameters: model, batch_idx
+    :type post_batch_callback: function
+    :param transform_to_device_fn: Function for sending data and labels to the
+                                   device. This provides an extensibility point
+                                   for performing any final transformations on
+                                   the data or targets, and determining what
+                                   actually needs to get sent to the device.
+    :type transform_to_device_fn: function
+    :param progress_bar: Optional :class:`tqdm` progress bar args.
+                         None for no progress bar
+    :type progress_bar: dict or None
+
+    :return: mean loss for epoch
+    :rtype: float
+    """
+    model.train()
+    # Use asynchronous GPU copies when the memory is pinned
+    # See https://pytorch.org/docs/master/notes/cuda.html
+    async_gpu = loader.pin_memory
+    if progress_bar is not None:
+        loader = tqdm(loader, **progress_bar)
+        # update progress bar total based on batches_in_epoch
+        if batches_in_epoch < len(loader):
```

Comment:
Here, you have defined your own custom train function. However, if you don't need some of these utilities (e.g., tqdm), you can delete the code and just say "Unused" for the corresponding parameter in the docstring.

See how we created a new and less-convoluted train function which is specific to dendrites, [here](https://github.com/numenta/nupic.research/blob/master/nupic/research/frameworks/dendrites/model_utils.py#L33-L50). If you delete a lot of these things that you know won't be used, that will seriously clean up the code here.

Reply:
The same goes for your eval function.

Reply:
Done